### ISSUE TITLE
Config option to have optional context keys

### ISSUE THREAD

**Issue by tomnez** [alexlafroscia/ember-context#43]
While using this addon in a project, I noticed that the `consume-context` helper (and `@context` decorator) will [throw an error](https://github.com/alexlafroscia/ember-context/blob/main/addon/-private/get-provider-for.js#L12) if the parent context does not exist.

The use case I was going for was something along the lines of styling a component differently if it was used inside the context of another specific component. Consider the following example where we may want to style a button differently if the button is inside a card component:

```html
<!-- templates/components/card.hbs -->
<ContextProvider @key="within-context" @value="card">
  {{yield}}
</ContextProvider>
```

```html
<!-- templates/components/button.hbs -->
<button
  class={{if (eq (consume-context "within-context") "card") "specificCardContextClass"}}
>
  {{@text}}
</button>
```

```html
<!-- main template -->
<Card>
  <p>Card stuff</p>

  <Button @text="Click me" />
</Card>
```

This example will currently error any time a `<Button>` is used _outside_ of a "card" context. I'm wondering if the `assert` could be changed to a `warn`, or if there could be an addon config option added where addon users could decide whether or not the parent contexts are always required. Something in `environment.js` like

```javascript
'ember-context': {
  requireParentContext: false,
}
```

or something along those lines? Then the helpers could check that option before erroring?

**Comment by alexlafroscia** [comment]
Hey @tomnez — that’s an interesting idea!

I’m a little hesitant to add the complexity of configuration options if they’re not totally necessary. Let’s see if we can work out a way to accomplish what you want without that!

Two ideas that I’m having right now:

1. Could you add a `ContextProvider` around your whole application that provides a “fallback” value? Something that matches on the same `key` but provides a value other than `”card”` in your case
2. What do you think about the individual usage of the context consumer providing a fallback? In that case, it would _not_ error if the provider was not found and would instead use the fallback value. Something like:

```hbs
{{! Returns `””` if the provider isn’t found }}
{{consume-context “within-context” fallback=“”}}
```

**Comment by tomnez** [comment]
Thanks @alexlafroscia - I would say option #2 with a fallback would be preferable and more flexible.

Reason being is that our intended use was to use `ember-context` in an addon, so we shouldn't expect or require that everyone using our addon wrap their application in a giant `ContextProvider`. This would be a good solution if we were just building our own standalone app, but the fact that it's an addon makes it a bit more difficult.

If you're good with this, is it something you'd like me to make a PR for?

**Comment by alexlafroscia** [comment]
> Reason being is that our intended use was to use ember-context in an addon, so we shouldn't expect or require that everyone using our addon wrap their application in a giant ContextProvider.

Totally makes sense! I like the fallback option.

>  is it something you'd like me to make a PR for?

That's always welcome!

**Comment by tomnez** [comment]
@alexlafroscia Hey Alex! I have a branch ready to PR, I'm getting authentication errors when trying to push my branch up to `origin`. Do I need to be added to the repo first?

**Comment by alexlafroscia** [comment]
Hmm, I don't think you should need that? Does it seem like an issue related to GitHub Actions? I've seen that before myself; replacing the `http` URL for `origin` with the `ssh` one fixes it for me (though I'm not really sure why...)

**Comment by tomnez** [comment]
I didn't fork, I followed the `CONTRIBUTING` guide so I just cloned it and made my own branch.

I tried setting `ssh` for `origin` and got permission denied for my username:

<img width="1130" alt="Screen Shot 2020-11-19 at 9 02 51 AM" src="https://user-images.githubusercontent.com/2292367/99684421-11867080-2a47-11eb-94a4-cc396ddcc455.png">

I'll keep poking around, this is likely something happening locally and nothing to do with the repo 🤔

**Comment by alexlafroscia** [comment]
Hmm, interesting point about the contribution docs! Thanks for bringing that up. I believe that they're basically the Ember default ones, which don't discuss the need to actually fork the repo to make a PR.

Generally the GitHub workflow is

1. Fork repo
2. Push branch to fork
3 Make PR from fork to original repo

If you fork the repo and push to that one, you should be able to get your branch up to GitHub and make a PR from there to here.

**Comment by alexlafroscia** [comment]
Closed by #46
